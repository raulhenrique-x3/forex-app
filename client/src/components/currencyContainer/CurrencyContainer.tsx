import { Avatar, Card, Container, Flex, Text } from "@chakra-ui/react";
import axios from "axios";
import React, { useState, useEffect } from "react";
import { BsArrowRightSquareFill } from "react-icons/bs";
import { IoLogoUsd } from "react-icons/io";
import { AiFillPoundCircle } from "react-icons/ai";
import { useNavigate } from "react-router-dom";
import io from "socket.io-client";

interface ICurrencyContainer {
  userId?: React.ReactNode;
  userName?: string;
  userEmail?: string;
  price?: any[];
  Value?: number;
  userWallet?: any;
  userBalance?: number;
  userGbpBalance?: number;
  showGBP?: boolean;
  showUSD?: boolean;
  showArrow?: boolean;
  onClick?: () => void;
}

const CurrencyContainer: React.FC<ICurrencyContainer> = ({ userId, showGBP, showUSD, showArrow }) => {
  const socket = io("ws://localhost:5000", { autoConnect: true });
  const [usdApi, setUsdApi] = useState<ICurrencyContainer[]>([]);
  const [gbpApi, setGbpApi] = useState<ICurrencyContainer[]>([]);
  const [ioResponse, setIoResponse] = useState<object[]>();
  const navigate = useNavigate();

  useEffect(() => {
    socket.on("Atualized data from API", (data) => {
      setIoResponse([data]);
    });
    setInterval(() => {
      socket.emit("Previous data from API");
    }, 1000);

    return () => {
      socket.off("Atualized data from API");
    };
  }, [socket]);

  useEffect(() => {
    axios
      .get(`http://localhost:5000/api/usd_to_gbp`)
      .then((res) => {
        setUsdApi([res?.data?.[0]]);
      })
      .catch((err) => console.error(err));
  }, [ioResponse]);

  useEffect(() => {
    axios
      .get(`http://localhost:5000/api/gbp_to_usd`)
      .then((res) => {
        setGbpApi([res?.data[0]]);
      })
      .catch((err) => console.error(err));
  }, [ioResponse]);

  return (
    <>
      {showUSD &&
        usdApi?.map((data, key) => (
          <Container
            key={key}
            marginTop={4}
            cursor={"pointer"}
            onClick={() => navigate(`/wallet/buy_currency/usd_to_gbp/${userId}`)}
          >
            <Card padding={4}>
              <Flex align={"center"} gap={8}>
                <Avatar bg="#ffffff" icon={<IoLogoUsd color="#000000" />} />
                <Text color={"#000000"} fontWeight={"bold"}>
                  USD-GBP: {data?.Value!}
                </Text>
                {showArrow && <BsArrowRightSquareFill color="#000000" cursor={"pointer"} />}
              </Flex>
            </Card>
          </Container>
        ))}
      {showGBP &&
        gbpApi?.map((data, key) => (
          <Container
            key={key}
            marginTop={4}
            cursor={"pointer"}
            onClick={() => navigate(`/wallet/buy_currency/gbp_to_usd/${userId}`)}
          >
            <Card padding={4}>
              <Flex alignItems={"center"} gap={8}>
                <Avatar bg="#ffffff" icon={<AiFillPoundCircle color="#000000" />} />
                <Text color={"#000000"} fontWeight={"bold"}>
                  GBP-USD: {data?.Value!}
                </Text>
                {showArrow && <BsArrowRightSquareFill color="#000000" cursor={"pointer"} />}
              </Flex>
            </Card>
          </Container>
        ))}
    </>
  );
};

export default CurrencyContainer;
