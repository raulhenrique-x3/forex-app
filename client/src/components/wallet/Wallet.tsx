import { Card, Container, Flex, Text } from "@chakra-ui/react";
import axios from "axios";
import { useEffect, useState } from "react";
import { useNavigate, useParams } from "react-router";
import { IUser } from "../../interface/interface";
import { BsFillArrowRightSquareFill } from "react-icons/bs";

interface IWallet {
  cursor?: "pointer";
  showArrow?: boolean;
}

export const Wallet: React.FC<IWallet> = ({ cursor, showArrow }) => {
  const [userData, setUserData] = useState<IUser[]>([]);
  let { userId } = useParams();
  const navigate = useNavigate();

  useEffect(() => {
    axios
      .get(`http://localhost:5000/home/${userId}`)
      .then((res) => {
        setUserData([res.data]);
      })
      .catch((error) => {
        console.error(error);
      });
  }, [userData]); // eslint-disable-line react-hooks/exhaustive-deps

  const navigateToWallet = () => {
    showArrow && navigate(`/wallet/${userId}`);
  };

  return (
    <Card
      padding={4}
      display={"flex"}
      backgroundColor={"#8236FD"}
      onClick={navigateToWallet}
      justifyContent={"space-around"}
      cursor={cursor}
      data-testid={"walletComponent"}
    >
      <Flex alignItems={"center"}>
        <Container>
          <Text fontSize={"lg"} color={"#ffffff"}>
            Wallet balance:{" "}
          </Text>
          <Text fontSize={"xl"} color={"#ffffff"} fontWeight={"bold"}>
            {new Intl.NumberFormat("en-US", {
              style: "currency",
              currency: "USD",
            }).format(userData[0]?.userWallet?.userBalance!)}
          </Text>
        </Container>
        {showArrow && <BsFillArrowRightSquareFill color={"#ffffff"} />}
      </Flex>
    </Card>
  );
};
