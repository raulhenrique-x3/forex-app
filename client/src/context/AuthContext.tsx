import React, { createContext, useMemo, useState } from "react";

interface IProps {
  isAuthenticated: boolean;
  login: (value: boolean) => void;
  logout: () => void;
}

export const AuthContext = createContext<IProps>({
  isAuthenticated: false,
  login() {},
  logout() {},
});

interface IChildren {
  children: React.ReactNode;
}

export function AuthContextProvider({ children }: IChildren) {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(false);

  const login = (value: boolean) => {
    setIsAuthenticated(value);
  };

  const logout = () => {
    setIsAuthenticated(false);
  };

  const value = useMemo(() => ({ isAuthenticated, login, logout }), [isAuthenticated]);
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export const AuthConsumer = AuthContext.Consumer;
