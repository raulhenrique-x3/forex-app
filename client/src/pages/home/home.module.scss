.desktopViewInfo {
  display: none;
}

@media (min-width: 768px) {
  .desktopViewInfo {
    display: block;
    width: 100%;
  }
}
