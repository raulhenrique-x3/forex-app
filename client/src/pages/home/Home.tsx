import { Button, Text, Box, Flex, Avatar, Heading, Input, Divider, Link, Container } from "@chakra-ui/react";
import axios from "axios";
import styles from "./home.module.scss";
import { useState, useEffect } from "react";
import { useParams } from "react-router";
import CurrencyContainer from "../../components/currencyContainer/CurrencyContainer";
import { DesktopHeaderInfos } from "../../components/desktopHeaderInfos/DesktopHeaderInfos";
import Header from "../../components/header/Header";
import { Wallet } from "../../components/wallet/Wallet";
import { IUser } from "../../interface/interface";

export const Home = () => {
  const [userData, setUserData] = useState<IUser[]>([]);
  let { userId } = useParams();
  useEffect(() => {
    axios
      .get(`http://localhost:5000/home/${userId}`)
      .then((res) => {
        setUserData([res.data]);
      })
      .catch((error) => {
        console.error(error);
      });
  }, []);
  return (
    <Flex w="100%" direction="column" alignItems="center" justifyContent="center" backgroundColor={"#F1F1F1"}>
      <Header user={userData} />
      <div className={styles.desktopViewInfo}>
        <DesktopHeaderInfos user={userData} />
      </div>
      <Container w="100%" height={"100vh"} maxWidth="400px" p={4}>
        <Box pt={4}>
          <Wallet />
        </Box>
        <Divider />
        <Container>
          <CurrencyContainer showGBP={true} showUSD={true} userId={userId} />
        </Container>
        <Divider />
      </Container>
    </Flex>
  );
};
