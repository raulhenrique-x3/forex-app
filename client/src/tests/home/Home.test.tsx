import { MemoryRouter, Route, Routes } from "react-router-dom";
import { AuthContextProvider } from "../../context/AuthContext";
import { Home } from "../../pages/home/Home";
import { cleanup, render, screen } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
import { create } from "react-test-renderer";
import axios from "../../__mocks__/axios";
import { Wallet } from "../../components/wallet/Wallet";
import CurrencyContainer from "../../components/currencyContainer/CurrencyContainer";
import "setimmediate";

jest.setTimeout(60000);
afterEach(cleanup);

describe("Access the home page", () => {
  it("should be have a wallet in home", async () => {
    create(
      <AuthContextProvider>
        <ChakraProvider>
          <MemoryRouter initialEntries={["/home/63e5135202704aea76e5fc52"]}>
            <Routes>
              <Route path={`/home/:id`} element={<Home />} />
            </Routes>
          </MemoryRouter>
        </ChakraProvider>
      </AuthContextProvider>
    );
    const wallet = screen.findByText("Wallet balance:");
    expect(wallet).toBeTruthy();
  });

  it("should to see your amount wallet fetched from api in home", async () => {
    axios.get.mockResolvedValueOnce({
      data: {
        userWallet: { usdAmount: 2000, gbpAmount: 1000, userBalance: 2000 },
        gbpUsd: { Value: 1.2095 },
        usdGbp: { Value: 0.8354 },
      },
    });

    render(
      <AuthContextProvider>
        <MemoryRouter initialEntries={["/home/63e5135202704aea76e5fc52"]}>
          <Wallet />
        </MemoryRouter>
      </AuthContextProvider>
    );

    const walletMoney = await screen.findByText("$2,000.00");
    expect(walletMoney).toBeTruthy();
  });

  it("should to see a currencyContainer with value fetched from api in home", async () => {
    axios.get.mockResolvedValueOnce({
      data: [
        {
          Value: 0.84,
        },
      ],
    });

    render(
      <AuthContextProvider>
        <ChakraProvider>
          <MemoryRouter initialEntries={["/home/63e5135202704aea76e5fc52"]}>
            <CurrencyContainer showUSD={true} />
          </MemoryRouter>
        </ChakraProvider>
      </AuthContextProvider>
    );

    const apiValue = await screen.findByText("USD-GBP: 0.84");
    expect(apiValue).toBeTruthy();
  });
});
