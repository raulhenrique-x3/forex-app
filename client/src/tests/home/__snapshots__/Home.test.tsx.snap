// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access the home page should be rendered 1`] = `
[
  <div
    className="css-122l52y"
  >
    <header
      className="header"
    >
      <a
        href="/home/undefined"
        onClick={[Function]}
      >
        <p
          className="headerText"
        >
          Forex-app
        </p>
      </a>
      <div
        className="desktopView"
      >
        <button
          className="chakra-button css-i99016"
          onClick={[Function]}
          type="button"
        >
          <svg
            color="#8236FD"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": "#8236FD",
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h16"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
          </svg>
        </button>
      </div>
      <nav
        className="desktopNav"
      >
        <ul
          className="desktopUl"
        >
          <li>
            <a
              href="/home/undefined"
              onClick={[Function]}
            >
              Home 
            </a>
          </li>
          <li>
            <a
              href="/wallet/undefined"
              onClick={[Function]}
            >
              Wallet 
            </a>
          </li>
          <li>
            <a
              href="/history/undefined"
              onClick={[Function]}
            >
              History 
            </a>
          </li>
        </ul>
        <button
          className="chakra-button css-jut409"
          onClick={[Function]}
          type="button"
        >
          Logout
        </button>
      </nav>
    </header>
    <div
      className="desktopViewInfo"
    >
      <div
        className="css-18vs6z"
      >
        <span
          className="chakra-avatar css-qlhcd1"
        >
          <svg
            aria-label=" avatar"
            className="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <div>
          <p
            className="chakra-text css-1xs2h3i"
          />
          <p
            className="chakra-text css-qqfgvy"
          />
          <button
            className="chakra-button css-tri09f"
            type="button"
          >
            ID: 
          </button>
        </div>
      </div>
    </div>
    <div
      className="chakra-container css-1oj9wzd"
    >
      <div
        className="css-1rwovhe"
      >
        <div
          className="chakra-card css-eeihmn"
          onClick={[Function]}
        >
          <div
            className="css-70qvj9"
          >
            <div
              className="chakra-container css-vhc7ah"
            >
              <p
                className="chakra-text css-12ukh6o"
              >
                Wallet balance:
                 
              </p>
              <p
                className="chakra-text css-proevc"
              >
                $NaN
              </p>
            </div>
            <svg
              color="#ffffff"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": "#ffffff",
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 14a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2a2 2 0 0 0-2 2v12zm4.5-6.5h5.793L8.146 5.354a.5.5 0 1 1 .708-.708l3 3a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708-.708L10.293 8.5H4.5a.5.5 0 0 1 0-1z"
              />
            </svg>
          </div>
        </div>
      </div>
      <hr
        aria-orientation="horizontal"
        className="chakra-divider css-svjswr"
      />
      <div
        className="chakra-container css-vhc7ah"
      />
      <hr
        aria-orientation="horizontal"
        className="chakra-divider css-svjswr"
      />
    </div>
  </div>,
  <span />,
  <span
    hidden={true}
    id="__chakra_env"
  />,
]
`;

exports[`Access the home page should to show a value of wallet in home 1`] = `
[
  <div
    className="css-122l52y"
  >
    <header
      className="header"
    >
      <a
        href="/home/undefined"
        onClick={[Function]}
      >
        <p
          className="headerText"
        >
          Forex-app
        </p>
      </a>
      <div
        className="desktopView"
      >
        <button
          className="chakra-button css-i99016"
          onClick={[Function]}
          type="button"
        >
          <svg
            color="#8236FD"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": "#8236FD",
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h16"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
          </svg>
        </button>
      </div>
      <nav
        className="desktopNav"
      >
        <ul
          className="desktopUl"
        >
          <li>
            <a
              href="/home/undefined"
              onClick={[Function]}
            >
              Home 
            </a>
          </li>
          <li>
            <a
              href="/wallet/undefined"
              onClick={[Function]}
            >
              Wallet 
            </a>
          </li>
          <li>
            <a
              href="/history/undefined"
              onClick={[Function]}
            >
              History 
            </a>
          </li>
        </ul>
        <button
          className="chakra-button css-jut409"
          onClick={[Function]}
          type="button"
        >
          Logout
        </button>
      </nav>
    </header>
    <div
      className="desktopViewInfo"
    >
      <div
        className="css-18vs6z"
      >
        <span
          className="chakra-avatar css-qlhcd1"
        >
          <svg
            aria-label=" avatar"
            className="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
        <div>
          <p
            className="chakra-text css-1xs2h3i"
          />
          <p
            className="chakra-text css-qqfgvy"
          />
          <button
            className="chakra-button css-tri09f"
            type="button"
          >
            ID: 
          </button>
        </div>
      </div>
    </div>
    <div
      className="chakra-container css-1oj9wzd"
    >
      <div
        className="css-1rwovhe"
      >
        <div
          className="chakra-card css-eeihmn"
          id="walletComponent"
          onClick={[Function]}
        >
          <div
            className="css-70qvj9"
          >
            <div
              className="chakra-container css-vhc7ah"
            >
              <p
                className="chakra-text css-12ukh6o"
              >
                Wallet balance:
                 
              </p>
              <p
                className="chakra-text css-proevc"
              >
                $NaN
              </p>
            </div>
            <svg
              color="#ffffff"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": "#ffffff",
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 14a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2a2 2 0 0 0-2 2v12zm4.5-6.5h5.793L8.146 5.354a.5.5 0 1 1 .708-.708l3 3a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708-.708L10.293 8.5H4.5a.5.5 0 0 1 0-1z"
              />
            </svg>
          </div>
        </div>
      </div>
      <hr
        aria-orientation="horizontal"
        className="chakra-divider css-svjswr"
      />
      <div
        className="chakra-container css-vhc7ah"
      />
      <hr
        aria-orientation="horizontal"
        className="chakra-divider css-svjswr"
      />
    </div>
  </div>,
  <span />,
  <span
    hidden={true}
    id="__chakra_env"
  />,
]
`;
